src/tomli/_parser.py

```diff
@@ -61,7 +62,30 @@ class TOMLDecodeError(ValueError):
     colno: The column corresponding to pos
     """
 
-    def __init__(self, msg: str, doc: str, pos: Pos):
+    def __init__(
+        self,
+        *args: Any,
+        msg: str | None = None,
```

Comment:
Shouldn't we only deprecate `TOMLDecodeError(msg, doc, pos)`? `TOMLDecodeError()` and `TOMLDecodeError(msg)` are fine.

Reply:
If we want to make progress towards a state where we have `json.JSONDecodeError` signature and where the exception's attributes always have a meaningful non-None value, then what you suggest is not enough. If we want an end state where passing exactly three positional arguments results in a special object similar to passing (msg, doc, pos) keyword arguments, then you're right.

Reply:
At the very least I would allow to pass the arguments positionally, this is too verbose:

```python
raise TOMLDecodeError(msg="Invalid statement", doc=src, pos=pos)
```

Reply:
Does verbosity matter if it stays within the library? Surely the "primary API" of this exception class is simply to get caught and have its attributes accessed, not being instantiated outside the library.

Reply:
Allowing arguments positionally results in a different `TOMLDecodeError("msg, "doc", 0).args` tuple so can theoretically break someone. Current state cannot.

Reply:
The only motivation for not allowing to pass them positionally is allowing to pass a 3-tuple to args.
That's even more obscure, so I don't think there's even a benefit to doing this.

Reply:
Sorry, could you clarify what you mean by "this" precisely? But yeah the motivation for many things in this PR is full backwards compatibility.

Reply:
Something like this:

```python
class TOMLDecodeError(ValueError):
    def __init__(self, msg: str | None = None, doc: str | None = None, pos: int | None = None):
        if msg is None or doc is None or pos is None:
            warn("not passing msg, doc or pos is deprecated", DeprecationWarning, 2)
            if msg is None:
                super().__init__()
            else:
                super().__init__(msg)
        else:
            ...
            self.msg = msg
            self.doc = doc
            self.pos = pos
            self.lineno = lineno
            self.colno = colno
```

Reply:
But wouldn't that
- completely break passing 4+ arguments
- break `exc.args` when passing 2 or 3 arguments
- break `exc.args` when passing one arg that is None

while current state breaks none of these cases.

Reply:
Yeah, in theory this not backwards compatible, but who would do this:

```python
>>> raise ValueError(1, 2, 3)
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    raise ValueError(1, 2, 3)
ValueError: (1, 2, 3)
```

Unless we know someone is relying on this, we shouldn't make the API significantly worse. If it does break someone, let them raise an issue.

Also, I'm pretty sure that deprecating passing the arguments positionally isn't what encukou meant.

It might be better to deprecate it in a followup PR.

Reply:
> Yeah, in theory this not backwards compatible, but who would do this:

Haha, don't ask me. I question why anyone would make an instance of TOMLDecodeError no matter what the args are they're using. But I can accept that in the stdlib we should expect anything.

It'd feel weird to me to have a strict back compatibility policy and then subjectively assess case by case whether it can be dodged.

> we shouldn't make the API significantly worse

IMO the API that really matters to users is being able to catch the exception and access attributes, which we don't make any worse. Creating instances we don't make any worse either, we just add a deprecation warning (we can also decide not to), in case someone creates in a way that doesn't add meaningful attribute values. But if we eventually want that all created instances have those attribute values, then deprecation is the way.


> If it does break someone, let them raise an issue.

Isn't this what we want to avoid with the back compat policy?

> It might be better to deprecate it in a followup PR.

This, or not deprecating at all, are fine by me. I think cpython codeowners should decide.

Reply:
> This, or not deprecating at all, are fine by me. I think cpython codeowners should decide.

Just create a PR, an then you'll know.